File: cdm/csharp/system/equipment/anesthesia_machine/SEAnesthesiaMachine.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/



public class SEAnesthesiaMachine // : SESystem
{
  // Keep enums in sync with appropriate schema/cdm/AnesthesiaMachineEnums.proto file !!
  public enum Connection : int
  {
    NullConnection = 0,/** Signals not provided, or no change */
    Off,
    Mask,
    Tube
  }
  extern string Connection_Name(Connection c);
  // Keep enums in sync with appropriate schema/cdm/AnesthesiaMachineEnums.proto file !!
  public enum OxygenSource : int
  {
    NullSource = 0,/** Signals not provided, or no change */
    NoSource,
    Wall,
    BottleOne,
    BottleTwo
  }
  extern string OxygenSource_Name(OxygenSource os);
  // Keep enums in sync with appropriate schema/cdm/AnesthesiaMachineEnums.proto file !!
  public enum PrimaryGas : int
  {
    NullGas = 0,/** Signals not provided, or no change */
    NoGas,
    Air,
    Nitrogen
  }
  extern string PrimaryGas_Name(PrimaryGas pg);

  protected Connection                        connection;
  protected SEScalarVolumePerTime             inletFlow;
  protected SEScalar                          inspiratoryExpiratoryRatio;
  protected SEScalar0To1                      oxygenFraction;
  protected OxygenSource                      oxygenSource;
  protected SEScalarPressure                  positiveEndExpiredPressure;
  protected PrimaryGas                        primaryGas;
  protected SEScalarFrequency                 respiratoryRate;
  protected SEScalarPressure                  reliefValvePressure;

  protected SEScalarPressure                  ventilatorPressure;

  //protected SEAnesthesiaMachineChamber        leftChamber;
  //protected SEAnesthesiaMachineChamber        rightChamber;

  //protected SEAnesthesiaMachineOxygenBottle   oxygenBottleOne;
  //protected SEAnesthesiaMachineOxygenBottle   oxygenBottleTwo;

  //protected Map<Event,Boolean> events = new HashMap<Event,Boolean>();

  public SEAnesthesiaMachine()
  {
    connection = Connection.NullConnection;
    inletFlow = null;
    inspiratoryExpiratoryRatio = null;
    oxygenFraction = null;
    oxygenSource = OxygenSource.NullSource;
    positiveEndExpiredPressure = null;
    primaryGas = PrimaryGas.NullGas;
    respiratoryRate = null;
    reliefValvePressure = null;

    ventilatorPressure = null;

    //leftChamber = null;
    //rightChamber = null;
    
    //oxygenBottleOne = null;
    //oxygenBottleTwo = null;
  }

  public void Clear()
  {
    connection = Connection.NullConnection;
    if (inletFlow != null)
      inletFlow.Invalidate();
    if (inspiratoryExpiratoryRatio != null)
      inspiratoryExpiratoryRatio.Invalidate();
    if (oxygenFraction != null)
      oxygenFraction.Invalidate();
    oxygenSource = OxygenSource.NullSource;
    if (positiveEndExpiredPressure != null)
      positiveEndExpiredPressure.Invalidate();
    primaryGas = PrimaryGas.NullGas;
    if (respiratoryRate != null)
      respiratoryRate.Invalidate();
    if (reliefValvePressure != null)
      reliefValvePressure.Invalidate();
    if (ventilatorPressure != null)
      ventilatorPressure.Invalidate();

    //if (HasLeftChamber())
    //  leftChamber.Clear();
    //if (HasRightChamber())
    //  rightChamber.Clear();
    //if (HasOxygenBottleOne())
    //  oxygenBottleOne.Clear();
    //if (HasOxygenBottleTwo())
    //  oxygenBottleTwo.Clear();

    //events.clear();
  }

  //public void Copy(SEAnesthesiaMachine from)
  //{
  //  Clear();
  //  if(from.connection!=null && from.connection != Connection.NullConnection)
  //  	this.connection=from.connection;      
  //  if(from.HasInletFlow())
  //    this.GetInletFlow().Set(from.GetInletFlow());
  //  if(from.HasInspiratoryExpiratoryRatio())
  //    this.GetInspiratoryExpiratoryRatio().Set(from.GetInspiratoryExpiratoryRatio());
  //  if(from.HasOxygenFraction())
  //    this.GetOxygenFraction().Set(from.GetOxygenFraction());
  //  if(from.oxygenSource!=null && from.oxygenSource != OxygenSource.NullSource)
  //  	this.setOxygenSource(from.GetOxygenSource());
  //  if(from.HasPositiveEndExpiredPressure())
  //    this.GetPositiveEndExpiredPressure().Set(from.GetPositiveEndExpiredPressure());
  //  if(from.primaryGas!=null && from.primaryGas != PrimaryGas.NullGas)
  //  	this.setPrimaryGas(from.GetPrimaryGas());
  //  if(from.HasReliefValvePressure())
  //    this.GetReliefValvePressure().Set(from.GetReliefValvePressure());
  //  if(from.HasRespiratoryRate())
  //    this.GetRespiratoryRate().Set(from.GetRespiratoryRate());
  //  if(from.HasVentilatorPressure())
  //    this.GetVentilatorPressure().Set(from.GetVentilatorPressure());    

  //  if(from.HasLeftChamber())    
  //    GetLeftChamber().copy(from.GetLeftChamber());
  //  if(from.HasRightChamber())
  //    GetRightChamber().copy(from.GetRightChamber());
  //  if(from.HasOxygenBottleOne())
  //    GetOxygenBottleOne().copy(from.GetOxygenBottleOne());
  //  if(from.HasOxygenBottleTwo())
  //    GetOxygenBottleTwo().copy(from.GetOxygenBottleTwo());

  //  events.putAll(from.events);
  //}

  //public void SetEvent(Event type, bool active)
  //{
  //  this.events.put(type, active);
  //}
  //public bool IsEventActive(Event type)
  //{
  //  if(!this.events.containsKey(type))
  //    return false;
  //  return this.events.Get(type);
  //}
  
  public Connection GetConnection()
  {
    return connection;
  }
  public void SetConnection(Connection c)
  {
    connection = c;
  }
  public bool HasConnection()
  {
    return connection != Connection.NullConnection;
  }

  public SEScalar GetInspiratoryExpiratoryRatio()
  {
    if (inspiratoryExpiratoryRatio == null)
      inspiratoryExpiratoryRatio = new SEScalar();
    return inspiratoryExpiratoryRatio;
  }
  public bool HasInspiratoryExpiratoryRatio()
  {
    return inspiratoryExpiratoryRatio == null ? false : inspiratoryExpiratoryRatio.IsValid();
  }

  public SEScalarVolumePerTime GetInletFlow()
  {
    if (inletFlow == null)
      inletFlow = new SEScalarVolumePerTime();
    return inletFlow;
  }
  public bool HasInletFlow()
  {
    return inletFlow == null ? false : inletFlow.IsValid();
  }

  public SEScalar0To1 GetOxygenFraction()
  {
    if (oxygenFraction == null)
      oxygenFraction = new SEScalar0To1();
    return oxygenFraction;
  }
  public bool HasOxygenFraction()
  {
    return oxygenFraction == null ? false : oxygenFraction.IsValid();
  }

  public OxygenSource GetOxygenSource()
  {
    return oxygenSource;
  }
  public void SetOxygenSource(OxygenSource os)
  {
  	oxygenSource = os;
  }
  public bool HasOxygenSource()
  {
    return oxygenSource != OxygenSource.NullSource;
  }

  public SEScalarPressure GetPositiveEndExpiredPressure()
  {
    if (positiveEndExpiredPressure == null)
      positiveEndExpiredPressure = new SEScalarPressure();
    return positiveEndExpiredPressure;
  }
  public bool HasPositiveEndExpiredPressure()
  {
    return positiveEndExpiredPressure == null ? false : positiveEndExpiredPressure.IsValid();
  }

  public PrimaryGas GetPrimaryGas()
  {
    return primaryGas;
  }
  public void SetPrimaryGas(PrimaryGas pg)
  {
  	primaryGas = pg;
  }
  public bool HasPrimaryGas()
  {
    return primaryGas != PrimaryGas.NullGas;
  }

  public SEScalarFrequency GetRespiratoryRate()
  {
    if (respiratoryRate == null)
      respiratoryRate = new SEScalarFrequency();
    return respiratoryRate;
  }
  public bool HasRespiratoryRate()
  {
    return respiratoryRate == null ? false : respiratoryRate.IsValid();
  }

  public SEScalarPressure GetReliefValvePressure()
  {
    if (reliefValvePressure == null)
      reliefValvePressure = new SEScalarPressure();
    return reliefValvePressure;
  }
  public bool HasReliefValvePressure()
  {
    return reliefValvePressure == null ? false : reliefValvePressure.IsValid();
  }
  
  public SEScalarPressure GetVentilatorPressure()
  {
    if (ventilatorPressure == null)
      ventilatorPressure = new SEScalarPressure();
    return ventilatorPressure;
  }
  public bool HasVentilatorPressure()
  {
    return ventilatorPressure == null ? false : ventilatorPressure.IsValid();
  }


  //public bool HasLeftChamber()
  //{
  //  return leftChamber == null ? false : true;
  //}
  //public SEAnesthesiaMachineChamber GetLeftChamber()
  //{
  //  if(leftChamber==null)
  //    this.leftChamber=new SEAnesthesiaMachineChamber();
  //  return leftChamber;
  //}
  //public void RemoveLeftChamber()
  //{
  //  leftChamber = null;
  //}

  //public bool HasRightChamber()
  //{
  //  return rightChamber == null ? false : true;
  //}
  //public SEAnesthesiaMachineChamber GetRightChamber()
  //{
  //  if(rightChamber==null)
  //    this.rightChamber=new SEAnesthesiaMachineChamber();
  //  return rightChamber;
  //}
  //public void RemoveRightChamber()
  //{
  //  rightChamber = null;
  //}

  //public SEAnesthesiaMachineOxygenBottle GetOxygenBottleOne()
  //{
  //  if (oxygenBottleOne == null)
  //    oxygenBottleOne = new SEAnesthesiaMachineOxygenBottle();
  //  return oxygenBottleOne;
  //}
  //public bool HasOxygenBottleOne()
  //{
  //  return oxygenBottleOne == null ? false : true;
  //}

  //public SEAnesthesiaMachineOxygenBottle GetOxygenBottleTwo()
  //{
  //  if (oxygenBottleTwo == null)
  //    oxygenBottleTwo = new SEAnesthesiaMachineOxygenBottle();

  //  return oxygenBottleTwo;
  //}
  //public bool HasOxygenBottleTwo()
  //{
  //  return oxygenBottleTwo == null ? false : true;
  //}
  
  public string ToString()
  {
    string leftChamber = "Left Chamber: NotProvided";
    //if(HasLeftChamber())
    //{
    //  leftChamber = GetLeftChamber().ToString();
    //  leftChamber = leftChamber.ReplaceAll("Chamber", "Left Chamber");
    //}
    string rightChamber = "Right Chamber: NotProvided";
    //if(HasRightChamber())
    //{
    //  rightChamber = GetRightChamber().ToString();
    //  rightChamber = rightChamber.ReplaceAll("Chamber", "Right Chamber");
    //}
    string o2BottleOne = "Oxygen Bottle One: NotProvided";
    //if(HasOxygenBottleOne())
    //{
    //  o2BottleOne = GetOxygenBottleOne().ToString();
    //  o2BottleOne = o2BottleOne.ReplaceAll("Bottle", "Bottle One");
    //}
    string o2BottleTwo = "Oxygen Bottle Two: NotProvided";
    //if(HasOxygenBottleTwo())
    //{
    //  o2BottleTwo = GetOxygenBottleTwo().ToString();
    //  o2BottleTwo = o2BottleTwo.ReplaceAll("Bottle", "Bottle Two");
    //}
    
      return "Anesthesia Machine"
      + "\n\tConnection: " + (HasConnection()? Connection_Name(GetConnection()):"NotProvided")
      + "\n\tInletFlow: " + GetInletFlow()
      + "\n\tInspiratoryExpiratoryRatio: " + GetInspiratoryExpiratoryRatio()
      + "\n\tOxygenFraction: " + GetOxygenFraction()
      + "\n\tOxygenSource: " + (HasOxygenSource()? OxygenSource_Name(GetOxygenSource()):"NotProvided")
      + "\n\tPositiveEndExpiredPressure: " + GetPositiveEndExpiredPressure()
      + "\n\tPrimaryGas: " + (HasPrimaryGas()? PrimaryGas_Name(GetPrimaryGas()):"NotProvided")
      + "\n\tRespiratoryRate: " + GetRespiratoryRate()
      + "\n\tReliefValvePressure: " + GetReliefValvePressure()
      + "\n\tVentilatorPressure: " + GetVentilatorPressure()
      + "\n\t" + leftChamber
      + "\n\t" + rightChamber
      + "\n\t" + o2BottleOne
      + "\n\t" + o2BottleTwo;
  }
}

File: cdm/cpp/system/equipment/anesthesiamachine/SEAnesthesiaMachine.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "system/equipment/anesthesiamachine/SEAnesthesiaMachine.h"
#include "system/equipment/anesthesiamachine/SEAnesthesiaMachineChamber.h"
#include "system/equipment/anesthesiamachine/SEAnesthesiaMachineOxygenBottle.h"
#include "substance/SESubstanceManager.h"

// State Actions
#include "system/equipment/anesthesiamachine/actions/SEAnesthesiaMachineConfiguration.h"

#include "properties/SEScalar0To1.h"
#include "properties/SEScalarPressure.h"
#include "properties/SEScalarFrequency.h"
#include "properties/SEScalarVolumePerTime.h"
#include "properties/SEScalarVolume.h"
#include "properties/SEScalarTime.h"
#include "io/protobuf/PBAnesthesiaMachine.h"

SEAnesthesiaMachine::SEAnesthesiaMachine(SESubstanceManager& substances) : SESystem(substances.GetLogger()), m_Substances(substances)
{
  m_Connection = eAnesthesiaMachine_Connection::NullConnection;
  m_InletFlow = nullptr;
  m_InspiratoryExpiratoryRatio = nullptr;
  m_OxygenFraction = nullptr;
  m_OxygenSource = eAnesthesiaMachine_OxygenSource::NullSource;
  m_PositiveEndExpiredPressure = nullptr;
  m_PrimaryGas = eAnesthesiaMachine_PrimaryGas::NullGas;  
  m_RespiratoryRate = nullptr;
  m_ReliefValvePressure = nullptr;
  m_VentilatorPressure = nullptr;
  m_LeftChamber = nullptr;
  m_RightChamber = nullptr;
  m_OxygenBottleOne = nullptr;
  m_OxygenBottleTwo = nullptr;
}

SEAnesthesiaMachine::~SEAnesthesiaMachine()
{
  Clear();
}

void SEAnesthesiaMachine::Clear()
{
  SESystem::Clear();

  m_Connection = eAnesthesiaMachine_Connection::NullConnection;
  SAFE_DELETE(m_InletFlow);
  SAFE_DELETE(m_InspiratoryExpiratoryRatio);
  SAFE_DELETE(m_OxygenFraction);
  m_OxygenSource = eAnesthesiaMachine_OxygenSource::NullSource;
  SAFE_DELETE(m_PositiveEndExpiredPressure);
  m_PrimaryGas = eAnesthesiaMachine_PrimaryGas::NullGas;
  SAFE_DELETE(m_RespiratoryRate);
  SAFE_DELETE(m_ReliefValvePressure);
  SAFE_DELETE(m_VentilatorPressure);
  SAFE_DELETE(m_LeftChamber);
  SAFE_DELETE(m_RightChamber);
  SAFE_DELETE(m_OxygenBottleOne);
  SAFE_DELETE(m_OxygenBottleTwo);
}

void SEAnesthesiaMachine::Merge(const SEAnesthesiaMachine& from)
{
  if(from.m_Connection!=eAnesthesiaMachine_Connection::NullConnection)
    SetConnection(from.m_Connection);
  COPY_PROPERTY(InletFlow);
  COPY_PROPERTY(InspiratoryExpiratoryRatio);
  COPY_PROPERTY(OxygenFraction);
  if (from.m_OxygenSource != eAnesthesiaMachine_OxygenSource::NullSource)
    SetOxygenSource(from.m_OxygenSource);
  COPY_PROPERTY(PositiveEndExpiredPressure);
  if (from.m_PrimaryGas != eAnesthesiaMachine_PrimaryGas::NullGas)
    SetPrimaryGas(from.m_PrimaryGas);
  COPY_PROPERTY(RespiratoryRate);
  COPY_PROPERTY(ReliefValvePressure);
  COPY_PROPERTY(VentilatorPressure);

  MERGE_CHILD(LeftChamber);
  MERGE_CHILD(RightChamber);
 
  MERGE_CHILD(OxygenBottleOne);
  MERGE_CHILD(OxygenBottleTwo);
}

bool SEAnesthesiaMachine::SerializeToString(std::string& output, SerializationFormat m) const
{
  return PBAnesthesiaMachine::SerializeToString(*this, output, m);
}
bool SEAnesthesiaMachine::SerializeToFile(const std::string& filename, SerializationFormat m) const
{
  return PBAnesthesiaMachine::SerializeToFile(*this, filename, m);
}
bool SEAnesthesiaMachine::SerializeFromString(const std::string& src, SerializationFormat m)
{
  return PBAnesthesiaMachine::SerializeFromString(src, *this, m);
}
bool SEAnesthesiaMachine::SerializeFromFile(const std::string& filename, SerializationFormat m)
{
  return PBAnesthesiaMachine::SerializeFromFile(filename, *this, m);
}

void SEAnesthesiaMachine::ProcessConfiguration(SEAnesthesiaMachineConfiguration& config)
{
  if (config.HasConfiguration())
    Merge(config.GetConfiguration());
  else if (config.HasConfigurationFile())
  {
    // Update the action with the file contents
    std::string cfg_file = config.GetConfigurationFile();
    if (!config.GetConfiguration().SerializeFromFile(cfg_file, JSON))
      Error("Unable to load configuration file", "SEAnesthesiaMachine::ProcessConfiguration");
    Merge(config.GetConfiguration());
  }
  StateChange();
}

const SEScalar* SEAnesthesiaMachine::GetScalar(const std::string& name)
{
  if (name == "InletFlow")
    return &GetInletFlow();
  if (name == "InspiratoryExpiratoryRatio")
    return &GetInspiratoryExpiratoryRatio();
  if (name == "OxygenFraction")
    return &GetOxygenFraction();
  if (name == "PositiveEndExpiredPressure")
    return &GetPositiveEndExpiredPressure();
  if (name == "ReliefValvePressure")
    return &GetReliefValvePressure();
  if (name == "RespiratoryRate")
    return &GetRespiratoryRate();
  if (name == "VentilatorPressure")
    return &GetVentilatorPressure();

  size_t split = name.find('-');
  if (split != name.npos)
  {
    std::string child = name.substr(0, split);
    std::string prop = name.substr(split + 1, name.npos);
    if (child == "LeftChamber")
      return GetLeftChamber().GetScalar(prop);
    if (child == "RightChamber")
      return GetRightChamber().GetScalar(prop);
    if (child == "OxygenBottleOne")
      return GetOxygenBottleOne().GetScalar(prop);
    if (child == "OxygenBottleTwo")
      return GetOxygenBottleTwo().GetScalar(prop);
  }

  return nullptr;
}

void SEAnesthesiaMachine::SetConnection(eAnesthesiaMachine_Connection c)
{
  m_Connection = c;
}
eAnesthesiaMachine_Connection SEAnesthesiaMachine::GetConnection() const
{
  return m_Connection;
}

bool SEAnesthesiaMachine::HasInletFlow() const
{
  return m_InletFlow == nullptr ? false : m_InletFlow->IsValid();
}
SEScalarVolumePerTime& SEAnesthesiaMachine::GetInletFlow()
{
  if (m_InletFlow == nullptr)
    m_InletFlow = new SEScalarVolumePerTime();
  return *m_InletFlow;
}
double SEAnesthesiaMachine::GetInletFlow(const VolumePerTimeUnit& unit) const
{
  if (m_InletFlow == nullptr)
    return SEScalar::dNaN();
  return m_InletFlow->GetValue(unit);
}

bool SEAnesthesiaMachine::HasInspiratoryExpiratoryRatio() const
{
  return m_InspiratoryExpiratoryRatio == nullptr ? false : m_InspiratoryExpiratoryRatio->IsValid();
}
SEScalar& SEAnesthesiaMachine::GetInspiratoryExpiratoryRatio()
{
  if (m_InspiratoryExpiratoryRatio == nullptr)
    m_InspiratoryExpiratoryRatio = new SEScalar();
  return *m_InspiratoryExpiratoryRatio;
}
double SEAnesthesiaMachine::GetInspiratoryExpiratoryRatio() const
{
  if (m_InspiratoryExpiratoryRatio == nullptr)
    return SEScalar::dNaN();
  return m_InspiratoryExpiratoryRatio->GetValue();
}

bool SEAnesthesiaMachine::HasOxygenFraction() const
{
  return m_OxygenFraction == nullptr ? false : m_OxygenFraction->IsValid();
}
SEScalar0To1& SEAnesthesiaMachine::GetOxygenFraction()
{
  if (m_OxygenFraction == nullptr)
    m_OxygenFraction = new SEScalar0To1();
  return *m_OxygenFraction;
}
double SEAnesthesiaMachine::GetOxygenFraction() const
{
  if (m_OxygenFraction == nullptr)
    return SEScalar::dNaN();
  return m_OxygenFraction->GetValue();
}

eAnesthesiaMachine_OxygenSource SEAnesthesiaMachine::GetOxygenSource() const
{
  return m_OxygenSource;
}
void SEAnesthesiaMachine::SetOxygenSource(eAnesthesiaMachine_OxygenSource src)
{
  m_OxygenSource = src;
}

bool SEAnesthesiaMachine::HasPositiveEndExpiredPressure() const
{
  return m_PositiveEndExpiredPressure == nullptr ? false : m_PositiveEndExpiredPressure->IsValid();
}
SEScalarPressure& SEAnesthesiaMachine::GetPositiveEndExpiredPressure()
{
  if (m_PositiveEndExpiredPressure == nullptr)
    m_PositiveEndExpiredPressure = new SEScalarPressure();
  return *m_PositiveEndExpiredPressure;
}
double SEAnesthesiaMachine::GetPositiveEndExpiredPressure(const PressureUnit& unit) const
{
  if (m_PositiveEndExpiredPressure == nullptr)
    return SEScalar::dNaN();
  return m_PositiveEndExpiredPressure->GetValue(unit);
}

eAnesthesiaMachine_PrimaryGas SEAnesthesiaMachine::GetPrimaryGas() const
{
  return m_PrimaryGas;
}
void SEAnesthesiaMachine::SetPrimaryGas(eAnesthesiaMachine_PrimaryGas gas)
{
  m_PrimaryGas = gas;
}

bool SEAnesthesiaMachine::HasRespiratoryRate() const
{
  return m_RespiratoryRate == nullptr ? false : m_RespiratoryRate->IsValid();
}
SEScalarFrequency& SEAnesthesiaMachine::GetRespiratoryRate()
{
  if (m_RespiratoryRate == nullptr)
    m_RespiratoryRate = new SEScalarFrequency();
  return *m_RespiratoryRate;
}
double SEAnesthesiaMachine::GetRespiratoryRate(const FrequencyUnit& unit) const
{
  if (m_RespiratoryRate == nullptr)
    return SEScalar::dNaN();
  return m_RespiratoryRate->GetValue(unit);
}

bool SEAnesthesiaMachine::HasReliefValvePressure() const
{
  return m_ReliefValvePressure == nullptr ? false : m_ReliefValvePressure->IsValid();
}
SEScalarPressure& SEAnesthesiaMachine::GetReliefValvePressure()
{
  if (m_ReliefValvePressure == nullptr)
    m_ReliefValvePressure = new SEScalarPressure();
  return *m_ReliefValvePressure;
}
double SEAnesthesiaMachine::GetReliefValvePressure(const PressureUnit& unit) const
{
  if (m_ReliefValvePressure == nullptr)
    return SEScalar::dNaN();
  return m_ReliefValvePressure->GetValue(unit);
}

bool SEAnesthesiaMachine::HasVentilatorPressure() const
{
  return m_VentilatorPressure == nullptr ? false : m_VentilatorPressure->IsValid();
}
SEScalarPressure& SEAnesthesiaMachine::GetVentilatorPressure()
{
  if (m_VentilatorPressure == nullptr)
    m_VentilatorPressure = new SEScalarPressure();
  return *m_VentilatorPressure;
}
double SEAnesthesiaMachine::GetVentilatorPressure(const PressureUnit& unit) const
{
  if (m_VentilatorPressure == nullptr)
    return SEScalar::dNaN();
  return m_VentilatorPressure->GetValue(unit);
}

bool SEAnesthesiaMachine::HasLeftChamber() const
{
  return m_LeftChamber == nullptr ? false : true;
}
SEAnesthesiaMachineChamber& SEAnesthesiaMachine::GetLeftChamber()
{
  if (m_LeftChamber == nullptr)
    m_LeftChamber = new SEAnesthesiaMachineChamber(m_Substances);
  return *m_LeftChamber;
}
const SEAnesthesiaMachineChamber* SEAnesthesiaMachine::GetLeftChamber() const
{
  return m_LeftChamber;
}
void SEAnesthesiaMachine::RemoveLeftChamber()
{
  SAFE_DELETE(m_LeftChamber);
}

bool SEAnesthesiaMachine::HasRightChamber() const
{
  return m_RightChamber == nullptr ? false : true;
}
SEAnesthesiaMachineChamber& SEAnesthesiaMachine::GetRightChamber()
{
  if (m_RightChamber == nullptr)
    m_RightChamber = new SEAnesthesiaMachineChamber(m_Substances);
  return *m_RightChamber;
}
const SEAnesthesiaMachineChamber* SEAnesthesiaMachine::GetRightChamber() const
{
  return m_RightChamber;
}
void SEAnesthesiaMachine::RemoveRightChamber()
{
  SAFE_DELETE(m_RightChamber);
}

bool SEAnesthesiaMachine::HasOxygenBottleOne() const
{
  return m_OxygenBottleOne == nullptr ? false : true;
}
SEAnesthesiaMachineOxygenBottle& SEAnesthesiaMachine::GetOxygenBottleOne()
{
  if (m_OxygenBottleOne == nullptr)
    m_OxygenBottleOne = new SEAnesthesiaMachineOxygenBottle(GetLogger());
  return *m_OxygenBottleOne;
}
const SEAnesthesiaMachineOxygenBottle* SEAnesthesiaMachine::GetOxygenBottleOne() const
{
  return m_OxygenBottleOne;
}
void SEAnesthesiaMachine::RemoveOxygenBottleOne()
{
  SAFE_DELETE(m_OxygenBottleOne);
}

bool SEAnesthesiaMachine::HasOxygenBottleTwo() const
{
  return m_OxygenBottleTwo == nullptr ? false : true;
}
SEAnesthesiaMachineOxygenBottle& SEAnesthesiaMachine::GetOxygenBottleTwo()
{
  if (m_OxygenBottleTwo == nullptr)
    m_OxygenBottleTwo = new SEAnesthesiaMachineOxygenBottle(GetLogger());
  return *m_OxygenBottleTwo;
}
const SEAnesthesiaMachineOxygenBottle* SEAnesthesiaMachine::GetOxygenBottleTwo() const
{
  return m_OxygenBottleTwo;
}
void SEAnesthesiaMachine::RemoveOxygenBottleTwo()
{
  SAFE_DELETE(m_OxygenBottleTwo);
}
